classes/migration/upgrade/v3_4_0/PreflightCheckMigration.php

```diff
@@ -32,752 +34,1144 @@ abstract protected function getContextTable(): string;
+     * @see https://github.com/pkp/pkp-lib/issues/7190
+     */
+    protected function checkSubmissionLocale(): void
+    {
+        // Fix missing submission.locale by retrieving the value from the publication
+        $rows = DB::table('submissions AS s')->join('publications AS p', 'p.publication_id', '=', 's.current_publication_id')
+            ->whereNull('s.locale')
+            ->whereNotNull('p.locale')
+            ->get(['s.submission_id', 'p.locale']);
+        foreach ($rows as $row) {
+            $this->_installer->log("Updating locale of submission {$row->submission_id} to {$row->locale}.");
+            DB::table('submissions')->where('submission_id', '=', $row->submission_id)->update(['locale' => $row->locale]);
+        }
+
+        if ($count = DB::table('submissions AS s')->whereNull('locale')->count()) {
+            throw new Exception("There are {$count} submission records with null in the locale column. Please correct these before upgrading.");
+        }
+    }
+
+    /**
+     * Checks if there are authors missing an user_group relationship
+     */
+    protected function checkAuthorsMissingUserGroup(): void
+    {
+        // Flag orphaned authors entries by user_group_id
+        $result = DB::table('authors AS a')->leftJoin('user_groups AS ug', 'ug.user_group_id', '=', 'a.user_group_id')->leftJoin('publications AS p', 'p.publication_id', '=', 'a.publication_id')->whereNull('ug.user_group_id')->select('a.author_id AS author_id', 'a.publication_id AS publication_id', 'a.user_group_id AS user_group_id', 'p.submission_id AS submission_id')->get();
+        foreach ($result as $row) {
+            $this->_installer->log("Found an orphaned author entry with author_id {$row->author_id} for publication_id {$row->publication_id} with submission_id {$row->submission_id} and user_group_id {$row->user_group_id}.");
+        }
+        if ($result->count()) {
+            throw new Exception('There are author records without matching user_group entries. Please correct these before upgrading.');
+        }
+    }
+
+    /**
+     * Clears orphaned entities before introducing foreign keys
+     * - The cleanup is executed based on the relationship dependencies between the entities
+     * - Some recovery might be attempted before cleaning/dropping data
+     * - Rows with required, but invalid foreign keys (null/bad values) will be deleted
+     * - Rows with nullable/optional foreign keys will be inspected on a case-by-case basis (if possible they will be nulled, otherwise removed)
+     * - Consideration is given to bidirectional/direct dependencies and exceptional cases (e.g. submission.current_publication_id, which is nullable, but required)
+     * @see https://github.com/pkp/pkp-lib/issues/6093
+     *
+     * @throws Exception
+     */
+    protected function clearOrphanedEntities(): void
+    {
+        $this->buildOrphanedEntityProcessor();
+        // Sort the tables by the number of dependent entities
```

Comment:
Not a big deal, but if we start deleting from entities that have a large number of dependents, it will decrease the number of things to re-check.